From 066659e113898fa9e78cf3b78be682d88ef04e3f Mon Sep 17 00:00:00 2001
From: Waqar Ahmed <waqar.17a@gmail.com>
Date: Fri, 1 Jul 2022 18:22:45 +0500
Subject: [PATCH] Allow configuring sidebar icon size

Since its now like the plasma panel, some people might want
a smaller button, some larger.
---
 apps/lib/kateconfigdialog.cpp | 19 +++++++++++++++++++
 apps/lib/kateconfigdialog.h   |  1 +
 apps/lib/katemdi.cpp          | 22 ++++++++++++++++++----
 apps/lib/katemdi.h            |  2 ++
 4 files changed, 40 insertions(+), 4 deletions(-)

diff --git a/apps/lib/kateconfigdialog.cpp b/apps/lib/kateconfigdialog.cpp
index fbc0e23c5..46a4643d4 100644
--- a/apps/lib/kateconfigdialog.cpp
+++ b/apps/lib/kateconfigdialog.cpp
@@ -159,6 +159,22 @@ void KateConfigDialog::addBehaviorPage()
         m_showTextForLeftRightSidebars->setChecked(cgGeneral.readEntry("Show text for left and right sidebar", false));
         connect(m_showTextForLeftRightSidebars, &QCheckBox::toggled, this, &KateConfigDialog::slotChanged);
         vbox->addWidget(m_showTextForLeftRightSidebars);
+
+        label = new QLabel(i18n("Icon size for left and right sidebar buttons"), buttonGroup);
+        m_leftRightSidebarsIconSize = new QSpinBox(buttonGroup);
+        m_leftRightSidebarsIconSize->setMinimum(16);
+        m_leftRightSidebarsIconSize->setMaximum(48);
+        m_leftRightSidebarsIconSize->setValue(cgGeneral.readEntry("Icon size for left and right sidebar buttons", 32));
+        connect(m_leftRightSidebarsIconSize, &QSpinBox::textChanged, this, &KateConfigDialog::slotChanged);
+        hlayout = new QHBoxLayout;
+        hlayout->addWidget(label);
+        hlayout->addWidget(m_leftRightSidebarsIconSize);
+        vbox->addLayout(hlayout);
+
+        connect(m_showTextForLeftRightSidebars, &QCheckBox::toggled, this, [l = QPointer(label), this](bool v) {
+            m_leftRightSidebarsIconSize->setEnabled(!v);
+            l->setEnabled(!v);
+        });
     }
 
     buttonGroup->setLayout(vbox);
@@ -433,6 +449,9 @@ void KateConfigDialog::slotApply()
         if (m_showTextForLeftRightSidebars) {
             cg.writeEntry("Show text for left and right sidebar", m_showTextForLeftRightSidebars->isChecked());
         }
+        if (m_leftRightSidebarsIconSize) {
+            cg.writeEntry("Icon size for left and right sidebar buttons", m_leftRightSidebarsIconSize->value());
+        }
 
         cg.writeEntry("Restore Window Configuration", sessionConfigUi.restoreVC->isChecked());
 
diff --git a/apps/lib/kateconfigdialog.h b/apps/lib/kateconfigdialog.h
index 4dedbc55d..25882025b 100644
--- a/apps/lib/kateconfigdialog.h
+++ b/apps/lib/kateconfigdialog.h
@@ -91,6 +91,7 @@ private:
     QComboBox *m_mouseForwardActions = nullptr;
     QCheckBox *m_modNotifications;
     QCheckBox *m_showTextForLeftRightSidebars = nullptr;
+    QSpinBox *m_leftRightSidebarsIconSize = nullptr;
     QComboBox *m_cmbQuickOpenListMode;
     QSpinBox *m_tabLimit;
     QCheckBox *m_autoHideTabs;
diff --git a/apps/lib/katemdi.cpp b/apps/lib/katemdi.cpp
index c179316d6..9ae01a323 100644
--- a/apps/lib/katemdi.cpp
+++ b/apps/lib/katemdi.cpp
@@ -333,15 +333,29 @@ Sidebar::Sidebar(KMultiTabBar::KMultiTabBarPosition pos, MainWindow *mainwin, QW
 
 void Sidebar::readConfig()
 {
+    bool needsUpdate = false;
+
     // shall we show text for the left and right bars?
     KSharedConfig::Ptr config = KSharedConfig::openConfig();
     KConfigGroup cgGeneral = KConfigGroup(config, "General");
     const bool showTextForLeftRight = cgGeneral.readEntry("Show text for left and right sidebar", false);
     if (showTextForLeftRight != m_showTextForLeftRight) {
         m_showTextForLeftRight = showTextForLeftRight;
-        for (const auto &tabs : m_idToWidget) {
-            updateButtonStyle(tab(tabs.first));
-        }
+        needsUpdate = true;
+    }
+
+    int size = cgGeneral.readEntry("Icon size for left and right sidebar buttons", 32);
+    if (size != m_leftRightSidebarIconSize) {
+        m_leftRightSidebarIconSize = size;
+        needsUpdate = true;
+    }
+
+    if (!needsUpdate) {
+        return;
+    }
+
+    for (const auto &tabs : m_idToWidget) {
+        updateButtonStyle(tab(tabs.first));
     }
 }
 
@@ -364,7 +378,7 @@ void Sidebar::updateButtonStyle(KMultiTabBarTab *button)
 {
     const auto originalText = button->property("kate_original_text").toString();
     if (!m_showTextForLeftRight && (position() == KMultiTabBar::Left || position() == KMultiTabBar::Right)) {
-        const int iconSize = style()->pixelMetric(QStyle::PM_LargeIconSize, nullptr, this);
+        const int iconSize = m_leftRightSidebarIconSize;
         button->setIconSize(QSize(iconSize, iconSize));
         button->setText(QString());
         button->setToolTip(originalText);
diff --git a/apps/lib/katemdi.h b/apps/lib/katemdi.h
index 4dff20d45..68df163ac 100644
--- a/apps/lib/katemdi.h
+++ b/apps/lib/katemdi.h
@@ -293,6 +293,8 @@ private:
 
     bool m_showTextForLeftRight = false;
 
+    int m_leftRightSidebarIconSize = 32;
+
 Q_SIGNALS:
     void sigShowPluginConfigPage(KTextEditor::Plugin *configpageinterface, int id);
 };
-- 
GitLab

